### PR TITLE
Issue 40621: Lineage update is not accurately represented in timeline event details

### PR DESCRIPTION
**Pull request by XingY**
#### Rationale
For LKSM timeline feature, we want to track lineage (data and material inputs) updates for samples when a sample is updated. Since data and material inputs are not added to samples table by default, those information are not available in the sample's row detail information. This pr will add data/material input fields to sample full row data. 

#### Related Pull Requests
* 

#### Changes
* added getNearestParentMaterials to ExperimentService
* Include material's data and material inputs in getRow

### CODE REVIEW

File: experiment/src/org/labkey/experiment/api/SampleSetUpdateServiceDI.java [hunk 1]
```diff
@@ -373,7 +382,55 @@ else if (lsid != null)
...
             throw new QueryUpdateServiceException("Either RowId or LSID is required to get Sample Set Material.");
 
-        return new TableSelector(getQueryTable(), filter, null).getMap();
+        Map<String, Object> sampleRow = new TableSelector(getQueryTable(), filter, null).getMap();
+
+        if (!addInputs)
+            return sampleRow;
+
+        ExperimentService experimentService = ExperimentService.get();
+        ExpMaterial seed = rowId != null ? experimentService.getExpMaterial(rowId) : experimentService.getExpMaterial(lsid);
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/api/SampleSetUpdateServiceDI.java:391]: Are we sure lsid is not null here?

File: experiment/src/org/labkey/experiment/api/SampleSetUpdateServiceDI.java [hunk 2]
```diff
@@ -373,7 +382,55 @@ else if (lsid != null)
...
+        Set<ExpMaterial> parentSamples = experimentService.getNearestParentMaterials(container, user, seed);
+        if (parentSamples != null && !parentSamples.isEmpty())
+            addParentFields(sampleRow, parentSamples, ExpMaterial.MATERIAL_INPUT_PARENT + "/", user);
+        Set<ExpData> parentDatas = experimentService.getNearestParentDatas(container, user, seed);
+        if (parentDatas != null && !parentDatas.isEmpty())
```
Comment (labkey-susanh) [inline experiment/src/org/labkey/experiment/api/SampleSetUpdateServiceDI.java]: It looks like `parentSamples` and `parentDatas` may always be non-null (could annotate `getNearestParent*` as such, so the `!= null` checks might be skipped.

**Review by labkey-susanh — APPROVED** [review]
Looks good.  Just a couple of suggestions.

File: experiment/src/org/labkey/experiment/api/SampleSetUpdateServiceDI.java (hunk 1 above)
Comment (XingY) [reply experiment/src/org/labkey/experiment/api/SampleSetUpdateServiceDI.java:391]: Exception would have been thrown if both are null

File: experiment/src/org/labkey/experiment/api/SampleSetUpdateServiceDI.java (hunk 2 above)
Comment (XingY) [reply experiment/src/org/labkey/experiment/api/SampleSetUpdateServiceDI.java]: Yep. Added NotNull and removed null check.

**Pull request merged**